### PR TITLE
Outfits may optionally reference another outfit into which they decompose

### PR DESCRIPTION
**Pull request by tehhowch**
Refs #4001 

 - If an outfit defines a `salvage` node, when viewing this outfit in the Boarding Panel, the option to salvage it is presented.
 - Attempting to salvage an outfit results in up to the specified maximum amount being added to the "available plunder" list, and shows a dialog with the plunder results.
   - The dialog uses the same format as that found when scanning a ship.
 - Salvage output can be `Outfit`s and commodities (e.g "Gold", "Uranium", "Heavy Metals", "Plastics").
   - If a commodity is specified that isn't defined in the trade system, it will not be available as salvage.
 - An optional attribute can be specified, which will associate that node's salvage with that attribute. If the boarding ship does not have that attribute, those salvage outputs will not be available.
 - Keywords `clear` and `remove` can be paired with an attribute name, to remove that particular attribute name if it is present for the outfit's salvage definition:
e.g. these are equivalent:
   - `salvage clear "mining lasers"`
   - `salvage remove "mining lasers"`
 - `salvage clear` and `salvage remove` (with no attribute) strip off the (default) no-attributes-required salvage output
 - The phrase `clear all` can be used to remove *every* salvage definition for the outfit:
`salvage clear all`

The current design has the "Salvage" option always available, even if the original outfit would fit in the player's cargo hold, as some plugins may define a "resource chain" in which the salvaged outfit is equally useful as the original outfit, presenting an interesting choice for the player.

A rework of the Boarding Panel art would most effectively integrate the "Salvage" option (i.e. widen the plunder list so the salvage option does not stick out a bit), but since the "Salvage" button will not even be displayed except when selecting an outfit that can be salvaged, the current piecemeal draw approach seems allowable. (Reworking the design of the `BoardingPanel` interface is beyond the scope of this PR.)

I expect this mechanic will offer some humorous and creative outlets for plugins, e.g. salvaging a jump drive into "Exotic Matter" or "Unknown" and so on. Possible uses in the base game include salvaging outfits into minables (Fission Reactor -> Uranium, etc.) or otherwise unobtainable outfits needed for missions ("we require parts from these 8 species hostile to us" -> player can negotiate for parts or steal & salvage).

<details><summary>Current UI (salvageable items are flagged even if you don't have space, since maybe they salvage into something you will have space for):</summary>

image
</details>

<details><summary>Dialog format (contents obviously for testing purposes):</summary>

image
(note the salvage has been added to the plunder list, in the background)
</details>

<details><summary>Salvage failure dialog:</summary>

image
</details>

<details><summary>Node syntax:</summary>

```
outfit <name>
    <normal outfit stuff>
    ...
    salvage clear all
    salvage (clear | remove) [<attribute name>]
    salvage [<attribute name>]
        (outfit | commodity) <name A> [<max count of A>]
        (outfit | commodity) <name B> [<max count of B>]
        ...
```
</details>

### CODE REVIEW

**Comment by Sageyyourpal** [comment]
I like very much, but I think you should need an outfit equipped for the salvage to be available, just an opinon

**Comment by tehhowch** [comment]
@Sageyyourpal  that's definitely a possibility that can be added. I don't plan to merge this for a while, in order to get a lot of feedback on it and the implementation, and suggestions such as that are definitely good feedback.
I could see adding an `attributes` node under `salvage` that would require said attribute to be present on the flagship in order to see salvage of that outfit as an option.

**Comment by Tadrix** [comment]
Two suggestions:

1. Let it be possible for them to be salvaged into commodities (most often into Electronics, Heavy Metals, Metals, and Plastic, most probably).
2. If this is possible, allow for the same to be done in the outfitter screen (because it's illogical to allow it only in space and not when you're in the workshop, and, with depreciation, sometimes it might be more useful to break up larger outfits into smaller than selling them and buying smaller ones).

**Comment by Zitchas** [comment]
Sounds like a pretty good thing. I like the potential of this. I second the idea for the them to give commodities. (actually, commodities should probably be the default thing that gets salvaged when there's nothing else suitable)

**Comment by jostephd** [comment]
I like the direction this is going. Defining the salvage explicitly allows it to be non-uniform, for example, a 10T power generator could define a 2T core worth 60% of the value and an 8T cooling outfit (that needn't even be available from outfitters the regular way).

Commodities are worth much less per ton than outfits so having them as fallback makes sense. It also aligns with @Sageyyourpal 's proposal, an alien species could be defined whose outfits can only be salvaged into scrap metal unless the player has acquired some special technology, or completed a mission string etc.

Besides Electronics, Heavy Metals, and Metals, the salvage could also have small amounts of pure metals (those harvested from asteroids)..

I wonder how ships protect themselves against their outfits being salvaged for parts. When you disable a pirate ship, suppose the captain of that ship is not his own boss but is leasing the ship from a superior pirate, how is that captain going to behave when his ship is disabled and is about to be looted for parts? There seem to be opportunities for additional gameplay there but of course it's beyond the scope of this PR. (The captain could offer to pay the player a mortgage in exchange for not being looted, or summon reinforcements, or defect to the player's side... and in the last case the _captain's boss_ could send two or three ships after him, to get him back dead or alive)

**Comment by Pointedstick** [comment]
This seems like it would enable some cool interactions. But so far it wouldn't be obvious to me looking at the boarding panel what the difference would be between "Salvage" and "Take". Why salvage something when you could just take it? Or is the idea that something that are or would be unplunderable would be salvageable instead? If that's the case, perhaps the button's text should change from "Take" to "Attempt salvage" or something like that.

**Comment by tehhowch** [comment]
> This seems like it would enable some cool interactions. But so far it wouldn't be obvious to me looking at the boarding panel what the difference would be between "Salvage" and "Take". Why salvage something when you could just take it?

The extra button is only shown for something salvageable, which means it will only appear when it can be used. "Take" to me is easy for players to understand ("I get this"), while I think "salvage" has the connotation of "I'm trying to get something of use from this"--my feeling may not be accurate. I think especially after boarding a few ships the meaning of "Take" will become quite clear. Perhaps "salvage" is not available for early outfits, such that its behavior being different will be clear.

One might salvage something for parts if they cannot fit the whole outfit into cargo (#4001 ), or if the player learns that the salvage components have other uses (e.g. outfit required for a mission, etc). 

> Or is the idea that something that are or would be unplunderable would be salvageable instead? If that's the case, perhaps the button's text should change from "Take" to "Attempt salvage" or something like that.

Currently there is nothing preventing one from combining `unplunderable` with `salvage`, though currently anything that is `unplunderable` is not shown in the boarding dialog if it is equipped - only if it is cargo:
https://github.com/endless-sky/endless-sky/blob/f8675db93be42c1e179b59291beaee41f37635f5/source/BoardingPanel.cpp#L88-L90

Whether it should be allowed to salvage an installed `unplunderable` or not is something that could be discussed (right now the plunder list doesn't remember where the item came from, so that would need to be amended to prevent taking an equipped `unplunderable`).

More text is not really an option unless we are fine with the button hanging off the side even more (it currently is juuust in the realm of "doesn't clash horrendously").

**Comment by MessyMix** [comment]
I don't think there would be any confusion if you could have a floating tooltip when you hover over the salvage button describing what it would do.

**Comment by Ferociousfeind** [comment]
Let me see if I understand this properly... I could conceivably make something like this and it would work "as intended"?
```
outfit "Triple Plasma Core"
    salvage "advanced salvaging"
        outfit "Plasma Core" 3
        commodity "Industrial" 1
        commodity "Electronics" 2
        commodity "Heavy Metals" 4
```
This would require your ship to have the "advanced salvaging" attribute (presumably either because it's a salvage ship or it has an outfit to the sound of "salvaging arm") which would break down a Triple Plasma Core into three Plasma Cores and some generic metals and things that held the three together.

**Comment by tehhowch** [comment]
@Ferociousfeind the counts are the possible maximums, not guaranteed results. So that definition could result in the player salvaging only a single Plasma Core and no commodities, or only commodities, etc. If you wanted a tiered approach, you could do something like

```bash
outfit "Triple Plasma Core"
    ...
    salvage
        commodity "Industrial" 1
        commodity "Electronics" 2
        commodity "Heavy Metals" 4
    salvage "basic salvaging"
        commodity "Industrial" 1
        commodity "Electronics" 2
        commodity "Heavy Metals" 4
    salvage "advanced salvaging"
        commodity "Industrial" 1
        commodity "Electronics" 2
        commodity "Heavy Metals" 4
        outfit "Plasma Core" 3
```
then the output result maximums would be like this:

|Flagship Attributes|Max "Industrial"|Max "Electronics"|Max "Heavy Metals"| Max "Plasma Cores"|
|:---|:---:|:---:|:---:|:---:|
|No special attributes|1|2|4|0|
|"basic salvaging"|2|4|8|0|
|"advanced salvaging"|2|4|8|3|
|"basic salvaging" & "advanced salvaging"|3|6|12|3|

**Comment by Ferociousfeind** [comment]
Oh, that is very interesting. I really like this PR simply because of all the new things you could DO! It opens just so much possibility, and could add more risk taking to plundering and salvaging. I'll have to look into this more, once it's merged. (If it's merged.)

**Review by Pointedstick — DISMISSED** [review]
All right, let's do it. More options never hurt content creators. :)

**Comment by Zitchas** [comment]
I really like the looks of this.

For testing the implementation of salvaging mods, the Remnant could be a good candidate. It fits their method of operation, and would limit the extent of salvaging stuff until the bugs get worked out.

**Comment by Ferociousfeind** [comment]
oh wait, could I guarantee you salvage three Plasma Cores by doing something like this?
```
outfit "Triple Plasma Core"
    ...
    salvage
        outfit "Plasma Core"
        commodity "Industrial"
    salvage
        outfit "Plasma Core"
        commodity "Electronics" 2
    salvage
        outfit "Plasma Core"
        commodity "Heavy Metals" 4
```
Or would that do not much, since there's still the chance you simply don't get anything?

**Comment by tehhowch** [comment]
@Ferociousfeind the salvage results are picked from the aggregated possible choices. Thus,

```c++
salvage "my attribute"
  outfit "A"
salvage "my attribute"
  outfit "A"
salvage "my attribute"
  outfit "A"
```
is equivalent to
```c++
salvage "my attribute"
  outfit "A" 3
```
For all possible item results, it is possible that a quantity of 0 is chosen. The distribution is uniform, so 0 is as likely as any other possible quantity.

I think I need to update this PR to condense the added quantities across salvage nodes; I believe the current implementation would list outfit A up to 3 times in the "salvage result" dialog.

**Comment by Ferociousfeind** [comment]
Would you be able to add some support for a minimum number of items salvaged? Ergo...
```
salvage [attribute]
    (outfit | commodity) <name> [<min#>] <max#>
```
this would guarantee at least <min#> of that outfit or commodity if two numbers are given, but would use the "default" behavior (min 0) if only one number specified.

**Comment by petervdmeer** [comment]
I like the suggestion of Ferocious to also allow setting a minimum number of items. This could also be useful if we want to present the salvage option in regular outfitters (for specific missions where you need to salvage things from alien artefacts).

File: source/Outfit.cpp [hunk 1]
```diff
@@ -201,6 +201,74 @@ void Outfit::Load(const DataNode &node)
...
 			attributes[child.Token(0)] = max(0., child.Value(1));
 		}
+		else if(child.Token(0) == "salvage")
+		{
+			// The associated salvage from this outfit is identified by a number
+			// of `salvage` nodes, each of which is associated with various outfits
+			// and/or commodities. An existing, attribute-specific salvage definition
+			// may be removed with the "remove" keyword: `salvage [[remove] <attribute>]`
+			// All existing salvage definitions may be erased with the "clear all"
+			// syntax: `salvage clear all`
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/Outfit.cpp:211]: Wait, I guess this is for when there are multiple salvage nodes on a single outfit; to allow later nodes (with complex attributes) to return more valuable items and clear the less valuable items from earlier nodes.

**Comment by petervdmeer** [comment]
I really like to see this feature added to the game, both in the boarding panel as well as in the Outfitter. (The second one for salvaging strange alien outfits from the players cargo at the outfitter.)
The Outfitter part could be kept for a later PR.

Would it make sense to also allow salvage definitions on unplunderable outfits? (Some outfits might be hard to remove from a ship, but you might still be able to break parts off of it.)

**Comment by Amazinite** [comment]
Closing all PRs older than 2020 which don't have any active discussion. If we wish to revisit these PRs then they may be reopened.

**Pull request closed without merging**